src/lib.rs

```diff
@@ -1,22 +1,101 @@
-      with an exception of `StorageClass::Handle` global variables.
-  3. Result of `Statement::Call` is automatically emitted.
-  4. `Statement::Emit` range is explicitly emitted.
+- [`Constant`]s and [`GlobalVariable`]s used by `EntryPoint`s and `Function`s, and
+
+- [`Type`]s used by the above.
+
+The body of an `EntryPoint` or `Function` is represented using two types:
+
+- An [`Expression`] produces a value, but has no side effects or control flow.
+  `Expressions` include variable references, unary and binary operators, and so
+  on.
+
+- A [`Statement`] can have side effects and structured control flow.
+  `Statement`s do not produce a value, other than by storing one in some
+  designated place. `Statements` include blocks, conditionals, and loops, but also
+  operations that have side effects, like stores and function calls.
+
+`Statement`s form a tree, with pointers into the DAG of `Expression`s.
+
+Restricting side effects to statements simplifies analysis and code generation.
+A Naga back end can generate code to evaluate an `Expression` however and
+whenever it pleases, as long as it is certain to observe the side effects of all
+previously executed `Statement`s.
+
+To improve translator performance and reduce memory usage, most structures are
+stored in an [`Arena`]. An `Arena<T>` stores a series of `T` values, indexed by
+[`Handle<T>`](Handle) values, which are just wrappers around integer indexes.
+For example, a `Function`'s expressions are stored in an `Arena<Expression>`,
+and compound expressions refer to their sub-expressions via `Handle<Expression>`
+values. (When examining the serialized form of a `Module`, note that the first
+element of an `Arena` has an index of 1, not 0.)
+
+## Function Calls
+
+The Naga IR's representation of function calls is unusual. Most languages treat
+function calls as expressions, but because calls may have side effects, Naga
+represents them with [`Statement::Call`]. A call statement may designate a
+particular `Expression` to represent its return value, if any, which can be used
+by subsequent statements and their expressions.
+
+## `Expression` evaluation time and scope
+
+While the order of execution of [`Statement`]s is apparent from their structure,
+it is not so obvious exactly when a given [`Expression`] should be evaluated,
+since many `Statement`s and `Expression`s may refer to its value. But it is
+essential to clearly specify an expression's evaluation time, since that
+determines which statements' effects the expression should observe. It is also
+helpful to back ends to limit the visibility of an `Expression`'s value to a
+portion of the statement tree.
+
+An `Expression` may only be used, whether by a `Statement` or another
+`Expression`, if one of the following is true:
+
+- The expression is an [`Expression::Constant`], [`Expression::FunctionArgument`], or
+  [`Expression::GlobalVariable`].
+
+- The expression is an [`Expression::LocalVariable`] that is either the
```

Comment:
Maybe we should just say "the expression type is a pointer" here and nothing else?

Reply:
At the moment, we don't document the types of any expressions, so the reader would have to go read the verifier to understand what "the expression type is a pointer" implies. But more broadly, the goal for the module-level documentation is to give the reader the orientation they need to make sense of the detailed documentation. So while it may be accurate to explain this in terms of what expressions generate pointers, using that approach here requires the reader to know details before they have an overview.